src/location/mod.rs

```diff
@@ -388,18 +434,41 @@ impl ObPartitionEntry {
         &self,
         part_id: i64,
     ) -> Option<&ObPartitionLocation> {
-        // logic_id = part_id in partition one
         self.parititon_location.get(&part_id)
     }
 
-    pub fn get_sub_partition_location_with_part_id(
+    pub fn get_partition_location_with_phy_id(
         &self,
-        part_id: i64,
+        phy_id: i64,
+        partid_tablet_map: Option<&HashMap<i64, i64>>,
+    ) -> Option<&ObPartitionLocation> {
+        // logic_id = part_id in partition one
+        let mut logic_id = phy_id;
+        if ob_vsn_major() >= 4 {
+            logic_id = partid_tablet_map.and_then(|m| m.get(&logic_id).copied())
+                .unwrap_or_else(|| {
+                    error!("get_sub_partition_location_with_part_id could not get tablet from logic id because the map is None or the logic_id is not present");
+                    -1
+                });
+        }
+        self.parititon_location.get(&logic_id)
+    }
+
+    pub fn get_sub_partition_location_with_phy_id(
+        &self,
+        phy_id: i64,
         sub_part_nums: i32,
+        partid_tablet_map: Option<&HashMap<i64, i64>>,
     ) -> Option<&ObPartitionLocation> {
-        // only for template sub partition
-        let logic_id = ob_part_constants::extract_part_idx(part_id) * sub_part_nums as i64
-            + ob_part_constants::extract_subpart_idx(part_id);
+        let mut logic_id = ob_part_constants::extract_part_idx(phy_id) * sub_part_nums as i64
+            + ob_part_constants::extract_subpart_idx(phy_id);
+        if ob_vsn_major() >= 4 {
+            logic_id = partid_tablet_map.and_then(|m| m.get(&logic_id).copied())
+                .unwrap_or_else(|| {
+                    error!("get_sub_partition_location_with_phy_id could not get tablet from logic id because the map is None or the logic_id is not present");
+                    -1
+                });
+        }
```

Comment:
```suggestion
        let mut logic_id = if ob_vsn_major() >= 4 {
            partid_tablet_map.and_then(|m| m.get(&logic_id).copied())
                .unwrap_or_else(|| {
                    error!("get_sub_partition_location_with_phy_id could not get tablet from logic id because the map is None or the logic_id is not present");
                    -1
                })
        } else {
            ob_part_constants::extract_part_idx(phy_id) * sub_part_nums as i64
            + ob_part_constants::extract_subpart_idx(phy_id);
        };
```

Reply:
Logic ID will be calculated and assigned before getting the real logic ID from the tablet map.
So if we use` let x = if {} else {}`, we need to write the same code twice, which is duplicated and useless.